front/components/data_source/DocumentOrTableUploadOrEditModal.tsx

```diff
@@ -0,0 +1,528 @@
+          } finally {
+            setUploading(false);
+          }
+        })();
+      }
+      setIsLoading(false);
+    } else {
+      resetTableOrDoc();
+    }
+  }, [isTable, initialId, table, owner.sId, dataSourceView.dataSource.name]);
+
+  const total = 0; // TODO: Get the total number of documents
+
+  if (
+    !initialId &&
+    plan.limits.dataSources.documents.count !== -1 &&
+    total >= plan.limits.dataSources.documents.count
+  ) {
+    return (
+      <DocumentLimitPopup
+        isOpen={isOpen}
+        plan={plan}
+        onClose={() => onClose(false)}
+        owner={owner}
+      />
+    );
+  }
+
+  const handleTableUpload = async () => {
+    setUploading(true);
+    try {
+      const fileContent = tableOrDoc.file
+        ? await handleFileUploadToText(tableOrDoc.file)
+        : null;
+      if (fileContent && fileContent.isErr()) {
+        return new Err(fileContent.error);
+      }
+
+      const csvContent = fileContent?.value
+        ? await parseAndStringifyCsv(fileContent.value.content)
+        : null;
+
+      if (csvContent && csvContent.length > 50_000_000) {
+        throw new Error("File too large");
+      }
+
+      const endpoint = `/api/w/${owner.sId}/data_sources/${dataSourceView.dataSource.name}/tables/csv`;
+      const body = JSON.stringify({
+        name: tableOrDoc.name,
+        description: tableOrDoc.description,
+        csv: undefined,
+        tableId: initialId,
+        timestamp: null,
+        tags: [],
+        parents: [],
+        truncate: false,
+        async: false,
+      });
+
+      const res = await fetch(endpoint, {
+        method: "POST",
+        headers: { "Content-Type": "application/json" },
+        body,
+      });
+
+      if (!res.ok) {
+        throw new Error("Failed to upsert table");
+      }
+
+      sendNotification({
+        type: "success",
+        title: `Table successfully ${initialId ? "updated" : "added"}`,
+        description: `Table ${tableOrDoc.name} was successfully ${initialId ? "updated" : "added"}.`,
+      });
+    } catch (error) {
+      sendNotification({
+        type: "error",
+        title: "Error upserting table",
+        description: `An error occurred: ${error instanceof Error ? error.message : String(error)}.`,
+      });
+    } finally {
+      setUploading(false);
+    }
+  };
+
+  const handleDocumentUpload = async () => {
+    setUploading(true);
+    try {
+      const endpoint = `/api/w/${owner.sId}/data_sources/${dataSourceView.dataSource.name}/documents/${encodeURIComponent(tableOrDoc.name)}`;
+      const body = JSON.stringify({
+        timestamp: null,
+        parents: null,
+        section: { prefix: null, content: tableOrDoc.text, sections: [] },
+        text: null,
+        source_url: tableOrDoc.sourceUrl || undefined,
+        tags: tableOrDoc.tags.filter(Boolean),
+        light_document_output: true,
+        upsert_context: null,
+        async: false,
+      } as PostDataSourceDocumentRequestBody);
```

Comment:
```suggestion
      const body: PostDataSourceDocumentRequestBody = {
        timestamp: null,
        parents: null,
        section: { prefix: null, content: tableOrDoc.text, sections: [] },
        text: null,
        source_url: tableOrDoc.sourceUrl || undefined,
        tags: tableOrDoc.tags.filter(Boolean),
        light_document_output: true,
        upsert_context: null,
        async: false,
      };
```

Then stringify when passing to the request's body.